compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,471 @@
+    // convention to use this encoding for cross-language LLVM CFI.
+    let encode_c_types = match fn_sig.abi {
+        // FIXME(rcvalle): See comment before the ty::Adt match in encode_ty().
+        Abi::C { .. } => false,
+        _ => false,
+    };
+
+    // Encode the return type
+    s.push_str(&encode_ty(tcx, fn_sig.output(), dict, encode_c_types));
+
+    // Encode the parameter types
+    let tys = fn_sig.inputs();
+    if !tys.is_empty() {
+        for ty in tys {
+            s.push_str(&encode_ty(tcx, *ty, dict, encode_c_types));
+        }
+
+        if fn_sig.c_variadic {
+            s.push('z');
+        }
+    } else {
+        if fn_sig.c_variadic {
+            s.push('z');
+        } else {
+            // Empty parameter lists, whether declared as () or conventionally as (void), are
+            // encoded with a void parameter specifier "v".
+            s.push('v')
+        }
+    }
+
+    // Close the "F..E" pair
+    s.push('E');
+
+    s.to_string()
+}
+
+/// Encodes a ty:Ty name, including its crate and path disambiguators and names.
+fn encode_ty_name<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> String {
+    let mut s = String::new();
+
+    // Start and namespace tags
+    let mut def_path = tcx.def_path(def_id);
+    def_path.data.reverse();
+    for disambiguated_data in &def_path.data {
+        s.push('N');
+        s.push_str(match disambiguated_data.data {
+            hir::definitions::DefPathData::TypeNs(_) => "t",
+            hir::definitions::DefPathData::ValueNs(_) => "v",
+            hir::definitions::DefPathData::ClosureExpr => "C",
+            hir::definitions::DefPathData::Ctor => "c",
+            hir::definitions::DefPathData::AnonConst => "k",
+            hir::definitions::DefPathData::ImplTrait => "i",
+            hir::definitions::DefPathData::CrateRoot
+            | hir::definitions::DefPathData::ForeignMod
+            | hir::definitions::DefPathData::Impl
+            | hir::definitions::DefPathData::LifetimeNs(_)
+            | hir::definitions::DefPathData::MacroNs(_)
+            | hir::definitions::DefPathData::Misc => {
+                bug!("encode_ty: unexpected `{:?}`", disambiguated_data.data);
+            }
+        });
+    }
+
+    // Crate disambiguator and name
+    s.push('C');
+    s.push_str(&to_disambiguator(tcx.stable_crate_id(def_path.krate).to_u64()));
+    let crate_name = format!("{}", tcx.crate_name(def_path.krate));
+    s.push_str(&format!("{}", crate_name.len()));
+    s.push_str(&crate_name);
+
+    // Disambiguators and names
+    def_path.data.reverse();
+    for disambiguated_data in &def_path.data {
+        let num = disambiguated_data.disambiguator as u64;
+        if num > 0 {
+            s.push_str(&to_disambiguator(num));
+        }
+        let name = format!("{}", disambiguated_data.data);
+        s.push_str(&format!("{}", name.len()));
+        s.push_str(&name);
+    }
+
+    s.to_string()
+}
```

Comment:
This isn't demangled or parsed at any time and also is the `<name>` part of `u<length><name>[I<element-type1..element-typeN>E]` as a vendor extended type, which would be fine if it were to be. Using a hash as `<name>` is as arbitrary as this[^1], wouldn't change/disambiguate/increase the uniqueness of the ids, but we'd lose control over granularity.

Additionally, if we were to use a hash, since it's a security feature, we'd need to have control over all or most aspects of it. I.e., what and how it's hashed, what hashing function is used, its collision resistance, and so on. I prefer to continue using this instead if it's okay for you.

What are the components that are missing possible substitutions so I can try to add it?

[^1]: Which is not exactly arbitrary; it implements v0's `<path>` for possible cases only to be used as the `<name>` part of `u<length><name>[I<element-type1..element-typeN>E]` as a vendor extended type.

Reply:
Oh, I missed that this is embedded inside `u <source-name>`. Thanks for correcting me.

If you would like to generate distinct identifiers for distinct types, while encoding the structure of Rust paths and Rust type system, it seems to me that you should expect something that is roughly comparable to v0 in overall complexity. I don't see any reason why it would be any simpler than that.

Reply:
It's simpler because the (1) types are monomorphized at the time they're encoded, (2) not every type may appear as a function parameter, and (3) `encode_ty_name` is not used for `ty::Ty` variants other than `ty::Adt`, `ty::FnDef`, and `ty::Dynamic`. So, `encode_ty_name` may safely implement encoding for these three variants only.

Reply:
One missing piece are generic arguments, which includes types in general as well as constants (e.g., `false`, `true`, `"abc"`, `Some(42)`, `[1, 2, 3]`, `S { a: 0, b: &13 }`, etc.). The scope is pretty much the same as in symbol mangling.

Reply:
Given that I'll fix `ty::Adt` encoding to encode substitutions instead of fields, will be anything missing? I don't see  substitutions for `ty::FnDef`, and `ty::Dynamic`. May those happen as well? May those also happen in path components of `ty::Adt`, `ty::FnDef`, and `ty::Dynamic`?

Reply:
I did some more research and it seems the answers for these questions are no. I'll resolve this conversation, but feel free to unresolve it if you have any more comments or suggestions.